k2/python/csrc/array.cc

```diff
@@ -0,0 +1,141 @@
+             type (i.e. `int32_t`), but may need type cast in other cases
+             (e.g. ValueType contains both `int32_t` and `float`).
+          d) cap_data.strides[0] == num-primitive-values in `ValueType`.
+          e) cap_data.strides[1] == 1.
+
+    Note if `data` in Array2 has stride > 1 (i.e. `data`'s type is
+    StridedPtr<ValueType>), the requirement of `cap_data` is nearly same with
+    case 2 above except cap_data.strides[0] will be greater than
+    num-primitive-values in `ValueType`.
+
+*/
+template <typename ValueType, bool IsPrimitive, typename I>
+class DLPackArray2;
+
+template <typename ValueType, typename I>
+class DLPackArray2<ValueType *, true, I> : public Array2<ValueType *, I> {
+ public:
+  DLPackArray2(py::capsule cap_indexes, py::capsule cap_data)
+      : indexes_tensor_(new Tensor(cap_indexes)),
+        data_tensor_(new Tensor(cap_data)) {
+    CHECK_EQ(indexes_tensor_->NumDim(), 1);
+    CHECK_GE(indexes_tensor_->Shape(0), 1);  // must have one element at least
+    CHECK_EQ(indexes_tensor_->Stride(0), 1);
+
+    CHECK_EQ(data_tensor_->NumDim(), 1);
+    CHECK_GE(data_tensor_->Shape(0), 0);  // num-elements
+    CHECK_EQ(data_tensor_->Stride(0), 1);
+
+    int32_t size1 = indexes_tensor_->Shape(0) - 1;
+    int32_t size2 = data_tensor_->Shape(0);
+    this->Init(size1, size2, indexes_tensor_->Data<I>(),
+               data_tensor_->Data<ValueType>());
+  }
+
+ private:
+  std::unique_ptr<Tensor> indexes_tensor_;
+  std::unique_ptr<Tensor> data_tensor_;
+};
+
+template <typename ValueType, typename I>
+class DLPackArray2<ValueType *, false, I> : public Array2<ValueType *, I> {
+ public:
+  DLPackArray2(py::capsule cap_indexes, py::capsule cap_data)
+      : indexes_tensor_(new Tensor(cap_indexes)),
+        data_tensor_(new Tensor(cap_data)) {
+    CHECK_EQ(indexes_tensor_->NumDim(), 1);
+    CHECK_GE(indexes_tensor_->Shape(0), 1);  // must have one element at least
+    CHECK_EQ(indexes_tensor_->Stride(0), 1);
+
+    CHECK_EQ(data_tensor_->NumDim(), 2);
+    CHECK_GE(data_tensor_->Shape(0), 0);  // num-elements
+    CHECK_EQ(data_tensor_->Shape(1) * data_tensor_->BytesPerElement(),
+             sizeof(ValueType));
+    CHECK_EQ(data_tensor_->Stride(0) * data_tensor_->BytesPerElement(),
+             sizeof(ValueType));
+    CHECK_EQ(data_tensor_->Stride(1), 1);
+
+    int32_t size1 = indexes_tensor_->Shape(0) - 1;
+    int32_t size2 = data_tensor_->Shape(0);
+    this->Init(size1, size2, indexes_tensor_->Data<I>(),
+               data_tensor_->Data<ValueType>());
+  }
+
+ private:
+  std::unique_ptr<Tensor> indexes_tensor_;
+  std::unique_ptr<Tensor> data_tensor_;
+};
+
+// Note: we can specialized for `StridedPtr` later if we need it.
+
+}  // namespace k2
+
+template <typename Ptr, bool IsPrimitive, typename I = int32_t>
+void PybindArray2Tpl(py::module &m, const char *name) {
+  using PyClass = k2::DLPackArray2<Ptr, IsPrimitive, I>;
+  using Parent = k2::Array2<Ptr, I>;
```

Comment:
I would suggest defining `Parent` inside  `DLPackArray2` and using

```cpp
using Parent = typename PyClass::Parent;
```

Reply:
what's the syntax you are writing? We can do this in C++?

Reply:
```cpp
template <typename ValueType, typename I>
class DLPackArray2<ValueType *, false, I> : public Array2<ValueType *, I> {
  public:
     using Parent = Array2<ValueType *, I>;
};

template <typename Ptr, bool IsPrimitive, typename I = int32_t>
void PybindArray2Tpl(py::module &m, const char *name) {
  using PyClass = k2::DLPackArray2<Ptr, IsPrimitive, I>;
  using Parent = typename PyClass::Parent;
};
```

Reply:
mm,  I prefer the original one as we only use it one time. Will resolve this issue